engine/src/witness/btc/smart_contract.rs

```diff
@@ -0,0 +1,338 @@
+use bitcoin::{
+	opcodes::all::{OP_PUSHDATA1, OP_RETURN},
+	ScriptBuf,
+};
+use cf_amm::common::{bounded_sqrt_price, sqrt_price_to_price};
+use cf_chains::{btc::ScriptPubkey, ForeignChainAddress};
+use cf_primitives::{Asset, AssetAmount, Price};
+use codec::{Decode, Encode};
+use itertools::Itertools;
+use utilities::SliceToArray;
+
+use crate::btc::rpc::VerboseTransaction;
+
+#[derive(PartialEq, Debug)]
+pub struct BtcContractCall {
+	output_asset: Asset,
+	deposit_amount: AssetAmount,
+	output_address: ForeignChainAddress,
+	// --- FoK ---
+	retry_duration: u16,
+	refund_address: ScriptPubkey,
+	min_price: Price,
+	// --- DCA ---
+	number_of_chunks: u16,
+	chunk_interval: u16,
+	// --- Boost ---
+	boost_fee: u8,
+}
+
+#[derive(Encode, Decode, Clone)]
+struct UtxoEncodedData {
+	output_asset: Asset,
+	output_address: ForeignChainAddress,
+	parameters: SharedCfParameters,
+}
+
+// The encoding of these parameters is the same across chains
+#[derive(Encode, Decode, Clone)]
+struct SharedCfParameters {
+	// FoK fields (refund address is stored externally):
+	retry_duration: u16,
+	min_output_amount: AssetAmount,
+	// DCA fields:
+	number_of_chunks: u16,
+	chunk_interval: u16,
+	// Boost fields:
+	boost_fee: u8,
+}
+
+#[allow(dead_code)]
+fn encode_data_in_nulldata_utxo(data: &[u8]) -> Option<ScriptBuf> {
+	const MAX_LEN: u8 = 80;
+
+	let len = data.len() as u8;
+	if len > MAX_LEN {
+		return None
+	}
+
+	let script_bytes: Vec<_> = [OP_RETURN.to_u8(), OP_PUSHDATA1.to_u8(), len]
```

Comment:
unfortunately its not that easy. have a look at `impl SerializeBtc for BitcoinOp`. If `len` is below 76, then we cant use OP_PUSHDATA1. instead we just push `len` followed by `data`. If len == 1 (will probably not happen), then we need special opcodes, too. At least we should chekc that len > 1 above.

Reply:
> yes. we will always push at least the destination address, so OP_PUSHDATA1 should work for all cases

This is what you told me in DM when I asked about it 🥲. Or did you mean that we would have to add padding for this to work? Why wouldn't `OP_PUSHDATA1` work for arbitrary length? I thought the special codes were just an optimisation... Seems silly of bitcoin not to support this.

> If len == 1 (will probably not happen), then we need special opcodes

Are you sure? `OP_PUSHBYTES_1` does exist and seems to work same as all other `OP_PUSHBYTES_x` opcodes.

Reply:
Updated now @martin-chainflip

Reply:
I would suggest we move the *encoding* of params (and associated types) into the chains crate. We already have utilities for pushing bytes that use the correct opcode depending on the size of the vec (see BitcoinOp::PushBytes). And the encoding will be used by the apis, which should not need to add a dependency on the engine witnessing code.

If we add the OP_RETURN variant to `BitcoinOp`, building the script will be something like `BitcoinScript::new([BitcoinOp::Return, BitcoinOp::PushBytes { params.encode() }])`.

Reply:
Moved encodeing to cf-chains @dandanlen (and now use `BitcoinScript`). The decoding is still in the engine since it should only be used by the witnesser. We can consider moving it too, but I think it might be worth waiting to see what the other components of btc smart contract will look like.